src/bin/pg_dump/dump_babel_utils.c

```diff
@@ -942,3 +944,179 @@ fixCopyCommand(Archive *fout, PQExpBuffer copyBuf, TableInfo *tbinfo, bool isFro
 	}
 	destroyPQExpBuffer(q);
 }
+
+/*
+ * hasSqlvariantColumn:
+ * Returns true if any of the columns in table is a sqlvariant data type column
+ */
+bool
+hasSqlvariantColumn(TableInfo *tbinfo)
+{
+	for (int i = 0; i < tbinfo->numatts; i++)
+		if (pg_strcasecmp(tbinfo->atttypnames[i],
+				quote_all_identifiers ? "\"sys\".\"sql_variant\"" : "sys.sql_variant") == 0)
+					return true;
+	return false;
+}
+
+/*
+ * getMbstrlen:
+ * returns the length of a multibyte string
+ */
+static int
+getMbstrlen(const char *mbstr, Archive *fout)
+{
+	int len = 0;
+	if (!mbstr)
+		return 0;
+	while (*mbstr){
+		mbstr += PQmblen(mbstr, fout->encoding);
+		len++;
+	}
+	return len;
+}
+
+/*
+ * fixCursorForBbfSqlvariantTableData:
+ * Prepare custom cursor for all Babelfish tables with atleast one sql_variant
+ * datatype column to correctly dump sql_variant data.
+ */
+int
+fixCursorForBbfSqlvariantTableData( Archive *fout,
+									TableInfo *tbinfo,
+									PQExpBuffer query,
+									int nfields,
+									int **sqlvar_metdata_pos)
+{
+	int orig_nfields = 0;
+	PQExpBuffer buf = createPQExpBuffer();
+
+	if (!isBabelfishDatabase(fout) || !hasSqlvariantColumn(tbinfo))
+		return 0;
+
+	*sqlvar_metdata_pos = (int *) pg_malloc0(3 * tbinfo->numatts * sizeof(int));
+	for (int i = 0; i < tbinfo->numatts; i++)
+	{
+		if (tbinfo->attisdropped[i])
+			continue;
+		if (tbinfo->attgenerated[i] && fout->dopt->column_inserts)
```

Comment:
We shouldn't be checking for `fout->dopt->column_inserts` since for COPY it won't be set but for tables with sql_variant columns we will assume this to be true.